jsonutil/puritan/puritan.go

```diff
@@ -0,0 +1,340 @@
+
+// Clean returns a slice of Clean()ed Strings.
+func (un SimpleStringSlice) Clean() []string {
+	out := make([]string, len(un))
+	for i := range un {
+		out[i] = un[i].Clean()
+	}
+
+	return out
+}
+
+// String accepts any valid JSON string
+type String struct {
+	s string
+}
+
+// NewString returns a new String whose Clean method returns the given string.
+//
+// This is meant for use in testing.
+func NewString(s string) String {
+	return String{s}
+}
+
+func (un *String) UnmarshalJSON(in []byte) (err error) {
+	un.s, err = unmarshal(in, uOpt{})
+	return
+}
+
+// Clean returns the string, with any control characters stripped.
+func (un String) Clean() string {
+	return un.s
+}
+
+// Paragraph accepts any valid JSON string
+type Paragraph struct {
+	s string
+}
+
+func (un *Paragraph) UnmarshalJSON(in []byte) (err error) {
+	un.s, err = unmarshal(in, uOpt{nlok: true})
+	return
+}
+
+// Clean returns the string, with any control characters except \n stripped.
+func (un Paragraph) Clean() string {
+	return un.s
+}
+
+// StringSlice is a slice of Strings, for convenience.
+type StringSlice []String
+
+// Clean returns a slice of Clean()ed Strings.
+func (un StringSlice) Clean() []string {
+	out := make([]string, len(un))
+	for i := range un {
+		out[i] = un[i].Clean()
+	}
+
+	return out
+}
+
+// OldPriceMap is a map for prices in the old API.
+type OldPriceMap struct {
+	s map[string]float64
+}
+
+func (un *OldPriceMap) UnmarshalJSON(in []byte) error {
+	var m map[string]float64
+	if err := json.Unmarshal(in, &m); err != nil {
+		return err
+	}
+	for k := range m {
+		if !okCurrency(k) {
+			return fmt.Errorf("invalid currency name %q", k)
+		}
+	}
+	un.s = m
+
+	return nil
+}
+
+// Clean returns a Cleaned map[string]float64.
+func (un OldPriceMap) Clean() map[string]float64 {
+	return un.s
+}
+
+// PriceMap is a map used for prices.
+//
+// the keys are ISO currency codes; the values are strings of floats.
+type PriceMap struct {
+	s map[string]string
+}
+
+func (un *PriceMap) UnmarshalJSON(in []byte) error {
+	var m map[string]string
+	if err := json.Unmarshal(in, &m); err != nil {
+		return err
+	}
+	for k, v := range m {
+		if !okCurrency(k) {
+			return fmt.Errorf("invalid currency name %q", k)
+		}
+		if !okPrice(v) {
+			return fmt.Errorf("invalid price %q", v)
+		}
+	}
+	un.s = m
+
+	return nil
+}
+
+// Clean returns a map[string]string of cleaned Strings.
+func (un PriceMap) Clean() map[string]string {
+	return un.s
+}
+
+func okCurrency(s string) bool {
+	if len(s) != 3 {
+		return false
+	}
+	for _, r := range s {
+		if r < 'A' || r > 'Z' {
+			return false
+		}
+	}
+	return true
+}
+
+func okPrice(s string) bool {
+	// limits to what strconv.ParseFloat accepts
+	for _, r := range s {
+		switch r {
+		case '.', '+', '-', 'e', 'E':
+			// ok
+		default:
+			if r < '0' || r > '9' {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func u4(in []byte) rune {
+	if len(in) < 6 || in[0] != '\\' || in[1] != 'u' {
+		return -1
```

Comment:
In the name of all the times our ancestors got this wrong in C :slightly_smiling_face: , can we please make this `return 0, false`?